File: opspilot-enterprise/packages/shared-types/src/memory-rag.ts
export type MemoryScope = "session" | "user" | "org" | "execution";
export type PiiLevel = "none" | "low" | "medium" | "high";

export interface MemoryUpsertRequest {
  tenant_id: string;
  scope: MemoryScope;
  subject_id: string;
  key: string;
  value_text: string;
  source_ref?: string;
  pii_level?: PiiLevel;
  retention_until?: string | null;
  metadata?: Record<string, unknown>;
  embedding?: number[] | null;
}

export interface MemoryUpsertResponse {
  memory_id: string;
  version_no: number;
  scope: MemoryScope;
  subject_id: string;
  key: string;
  created_at: string;
  storage_backend: "sqlite" | "postgres";
}

export interface RagRetrieveRequest {
  run_id?: string | null;
  tenant_id: string;
  query: string;
  top_k?: number;
  scopes?: MemoryScope[];
  environment?: string | null;
  object_type?: string | null;
  object_id?: string | null;
}

export interface RagHit {
  ref_id: string;
  source_type: string;
  title: string;
  summary: string;
  score: number;
  source?: string;
  metadata?: Record<string, unknown>;
}

export interface RagRetrieveResponse {
  query: string;
  hits: RagHit[];
  citations: Array<Record<string, unknown>>;
  evidence_refs: string[];
  insufficient_evidence: boolean;
  reason: string;
}

File: packages/shared-schema/src/opspilot_schema/memory_rag.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

MemoryScope = Literal["session", "user", "org", "execution"]
PiiLevel = Literal["none", "low", "medium", "high"]


class MemoryUpsertRequest(BaseModel):
    tenant_id: str
    scope: MemoryScope
    subject_id: str
    key: str
    value_text: str
    source_ref: str = ""
    pii_level: PiiLevel = "none"
    retention_until: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    embedding: list[float] | None = None


class MemoryUpsertResponse(BaseModel):
    memory_id: str
    version_no: int
    scope: MemoryScope
    subject_id: str
    key: str
    created_at: str
    storage_backend: Literal["sqlite", "postgres"] = "sqlite"


class RagRetrieveRequest(BaseModel):
    run_id: str | None = None
    tenant_id: str
    query: str
    top_k: int = 5
    scopes: list[MemoryScope] = Field(default_factory=lambda: ["org", "user", "session"])
    environment: str | None = None
    object_type: str | None = None
    object_id: str | None = None


class RagHit(BaseModel):
    ref_id: str
    source_type: str
    title: str
    summary: str
    score: float
    source: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)


class RagRetrieveResponse(BaseModel):
    query: str
    hits: list[RagHit] = Field(default_factory=list)
    citations: list[dict[str, Any]] = Field(default_factory=list)
    evidence_refs: list[str] = Field(default_factory=list)
    insufficient_evidence: bool = False
    reason: str = ""
